libcst/tool.py

```diff
@@ -44,6 +44,98 @@
+
+    Setting ``show_whitespace`` to ``False`` will filter whitespace fields.
+
+    Setting ``show_defaults`` to ``False`` will filter fields if their value is equal to
+    the default value ;  while respecting  the value of ``show_whitespace``.
+
+    Setting ``show_syntax``  to ``False`` will filter syntax fields ; while respecting
+    the value of ``show_whitespace`` & ``show_defaults``.
+    """
+
+    fields: Sequence["dataclasses.Field[CSTNode]"] = dataclasses.fields(node)
+
+    # Hide all fields prefixed with "_"
+    fields = [f for f in fields if f.name[0] != "_"]
+
+    # Filter whitespace nodes if needed
+    if not show_whitespace:
+
+        def _is_whitespace(field: "dataclasses.Field[object]") -> bool:
+            if "whitespace" in field.name:
+                return True
+            if "leading_lines" in field.name:
+                return True
+            if "lines_after_decorators" in field.name:
+                return True
+            if isinstance(node, (IndentedBlock, Module)) and field.name in [
+                "header",
+                "footer",
+            ]:
+                return True
+            if isinstance(node, IndentedBlock) and field.name == "indent":
+                return True
+            return False
+
+        fields = [f for f in fields if not _is_whitespace(f)]
+    # Filter values which aren't changed from their defaults
+    if not show_defaults:
+
+        def _get_default(fld: "dataclasses.Field[object]") -> object:
+            if fld.default_factory is not dataclasses.MISSING:
+                # pyre-fixme[29]: `Union[dataclasses._MISSING_TYPE,
+                #  dataclasses._DefaultFactory[object]]` is not a function.
+                return fld.default_factory()
+            return fld.default
+
+        fields = [
+            f
+            for f in fields
+            if not deep_equals(getattr(node, f.name), _get_default(f))
+        ]
+    # Filter out values which aren't interesting if needed
+    if not show_syntax:
+
+        def _is_syntax(field: "dataclasses.Field[object]") -> bool:
+            if isinstance(node, Module) and field.name in [
+                "encoding",
+                "default_indent",
+                "default_newline",
+                "has_trailing_newline",
+            ]:
+                return True
+            type_str = repr(field.type)
+            if (
+                "Sentinel" in type_str
+                and field.name not in ["star_arg", "star", "posonly_ind"]
+                and "whitespace" not in field.name
+            ):
+                # This is a value that can optionally be specified, so its
+                # definitely syntax.
+                return True
+
+            for name in ["Semicolon", "Colon", "Comma", "Dot", "AssignEqual"]:
+                # These are all nodes that exist for separation syntax
+                if name in type_str:
+                    return True
+
+            return False
```

Comment:
This could be its own public helper function

Reply:
Following functions - with docstring and tests - have been created in ``libcst.helpers``:
- ``get_node_fields``
- ``is_whitespace_node_field``
- ``is_syntax_node_field``
- ``get_field_default_value``
- ``is_default_node_field``
- ``filter_node_fields``  

All above functions are now part the Sphinx Helpers doc.

``libcst.tool.dump`` now uses ``filter_node_fields``.

Reply:
*(Please check the tests ; I was not really sure on how to properly cover use cases, so do tell if works need to be done there)*